src/completion_items_cache.ts

```diff
@@ -15,20 +17,65 @@ export function refresh() {
         console.warn("No workspace folder. typescript-namespace-imports-vscode-plugin will not work");
         return;
     }
-    workspaceCompletionItems = workspaceFolders.map(workspaceFolder => new vscode.CompletionList([], true));
+    workspaceCompletionItems = workspaceFolders.map(_ => new vscode.CompletionList([], true));
     workspaceFolders.forEach(workspaceFolder => {
-        const relativePattern = new vscode.RelativePattern(workspaceFolder, "**/*.ts");
-        vscode.workspace.findFiles(relativePattern).then(
-            uris => {
-                const completionItems = uris.map(uri => uriToCompletionItem(uri));
-                workspaceCompletionItems[workspaceFolder.index] = new vscode.CompletionList(completionItems, false);
-            },
-            error => {
-                console.error("Error in updateWorkspaceCompletionItems: " + error);
+        const tsconfigPattern = new vscode.RelativePattern(workspaceFolder, "**/tsconfig.json");
+        const typescriptPattern = new vscode.RelativePattern(workspaceFolder, "**/*.ts");
+        vscode.workspace
+            .findFiles(tsconfigPattern)
+            .then(_tsconfigUrisToBaseUrlMap(workspaceFolder), error => {
+                console.error(`Error while working with **/tsconfig.json files ${error}`);
+                return {};
+            })
+            .then(baseUrlMap => {
+                vscode.workspace.findFiles(typescriptPattern).then(
+                    uris => {
+                        const completionItems = uris.map(uri => uriToCompletionItem(uri, baseUrlMap));
+                        workspaceCompletionItems[workspaceFolder.index] = new vscode.CompletionList(
+                            completionItems,
+                            false
+                        );
+                    },
+                    error => {
+                        console.error(`Error in updateWorkspaceCompletionItems: ${error}`);
+                    }
+                );
             });
     });
 }
 
+const _tsconfigDocumentToBaseUrl = (tsconfigDoc: vscode.TextDocument): string | null => {
+    const tsconfigObj = JSON.parse(stripJsonComments(tsconfigDoc.getText()));
+    if ("compilerOptions" in tsconfigObj && "baseUrl" in tsconfigObj["compilerOptions"]) {
+        return <string>tsconfigObj["compilerOptions"]["baseUrl"];
+    }
+    return null;
+};
+
+const _tsconfigUrisToBaseUrlMap = (workspaceFolder: vscode.WorkspaceFolder) => (
+    uris: vscode.Uri[]
+): Thenable<Record<string, string>> => {
+    const recordPromises = Promise.all(
+        uris.map(tsconfigUri =>
+            vscode.workspace
+                .openTextDocument(tsconfigUri)
+                .then(_tsconfigDocumentToBaseUrl)
```

Comment:
Is this one a promise at all? I think i would avoid chaining it if not. You could make the following then take in a file and then just call `_tsconfigDocumentToBaseUrl(file)` like normal synchronous code

Reply:
it doesn't emit `Promise` itself, but works with a code inside `Promise`